src/main/java/fr/xephi/authme/service/BungeeService.java

```diff
@@ -1,69 +1,191 @@
     private BukkitService service;
 
     private boolean isEnabled;
-    private String bungeeServer;
+    private String destinationServerOnLogin;
+
+    private DataSource dataSource;
 
     /*
      * Constructor.
      */
     @Inject
-    BungeeService(AuthMe plugin, BukkitService service, Settings settings) {
+    BungeeService(AuthMe plugin, BukkitService service, Settings settings, DataSource dataSource) {
         this.plugin = plugin;
         this.service = service;
+        this.dataSource = dataSource;
         reload(settings);
     }
 
+    @Override
+    public void reload(Settings settings) {
+        this.isEnabled = settings.getProperty(HooksSettings.BUNGEECORD);
+        this.destinationServerOnLogin = settings.getProperty(HooksSettings.BUNGEECORD_SERVER);
+        Messenger messenger = plugin.getServer().getMessenger();
+        if (!this.isEnabled) {
+            return;
+        }
+        if (!messenger.isOutgoingChannelRegistered(plugin, "BungeeCord")) {
+            messenger.registerOutgoingPluginChannel(plugin, "BungeeCord");
+        }
+        if (!messenger.isIncomingChannelRegistered(plugin, "BungeeCord")) {
+            messenger.registerIncomingPluginChannel(plugin, "BungeeCord", this);
+        }
+    }
+
+    private void sendBungeecordMessage(String... data) {
+        ByteArrayDataOutput out = ByteStreams.newDataOutput();
+        for(String element : data) {
+            out.writeUTF(element);
+        }
+        service.sendPluginMessage("BungeeCord", out.toByteArray());
+    }
+
     /**
      * Send a player to a specified server. If no server is configured, this will
      * do nothing.
      *
      * @param player The player to send.
      */
-    public void connectPlayer(Player player) {
-        if (!isEnabled || bungeeServer.isEmpty()) {
+    public void connectPlayerOnLogin(Player player) {
+        if (!isEnabled || destinationServerOnLogin.isEmpty()) {
             return;
         }
 
-        service.scheduleSyncDelayedTask(new Runnable() {
-            @Override
-            public void run() {
-                ByteArrayDataOutput out = ByteStreams.newDataOutput();
-                out.writeUTF("Connect");
-                out.writeUTF(bungeeServer);
-                player.sendPluginMessage(plugin, "BungeeCord", out.toByteArray());
-            }
-        }, 20L);
+        service.scheduleSyncDelayedTask(() ->
+            sendBungeecordMessage("Connect", player.getName(), destinationServerOnLogin), 20L);
+    }
+
+    private void sendAuthMeBungeecordMessage(String type, String... data) {
+        if(!isEnabled) {
+            return;
+        }
+
+        List<String> dataList = Arrays.asList(data);
+        dataList.add(0, "AuthMe");
+        dataList.add(1, type);
+        sendBungeecordMessage(dataList.toArray(new String[dataList.size()]));
+    }
+
+    public void sendLogin(String name) {
+        sendAuthMeBungeecordMessage(AuthMeBungeeMessageType.LOGIN, name.toLowerCase());
```

Comment:
Up to you, just putting the idea out there: does it make sense to have dedicated methods for all of these cases if we could expose the `AuthMeBungeeMessageType` as an enum instead and have that coming in as an argument? You'd save on a lot of those specific methods and could easily add new entries to that enum without requiring a method counterpart to be added, too.

Reply:
That's because we might have some messages with extra arguments in the future, idk how to handle that correctly, any idea?

Reply:
Personally, I'd handle that when the need for it arises. Like, you could have a `void sendMessage(AuthMeBungeeMessageType)` and later on the first time it comes up, you can easily overload it with a `void sendMessage(AuthMeBungeeMessageType, String... args)` method or whatever.